### PR TITLE
APP-464: [C++ SDK] GPIO control slave application

### PR DESCRIPTION
**Pull request by vasner**
Add CPP posix demo source code;

Review: @Slash32 and @rasendubi

### CODE REVIEW

**Comment by kaa-leeroy** [comment]
Can one of the admins verify this patch?

File: gpiocontrol/source/cpp/src/KaaDemo.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            setLedStatus(event.gpio.id, true);
+        } else {
+            setLedStatus(event.gpio.id, false);
+        }
+        
+        printLedStatus();
+    }
+
+protected:
+    RemoteControlECF &remote;
```
Comment (RostakaGmfun) [inline gpiocontrol/source/cpp/src/KaaDemo.cpp:71]: Please append underscore `_` to private members

File: gpiocontrol/source/cpp/src/KaaDemo.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    std::vector<nsRemoteControlECF::GpioStatus> leds;
+
+private:
+    void setLedStatus(int id, bool status)
+    {
+        for (std::size_t i = 0; i < leds.size(); ++i) {
```
Comment (RostakaGmfun) [inline gpiocontrol/source/cpp/src/KaaDemo.cpp]: Please prefer range-based for loop in C++11

File: gpiocontrol/source/cpp/src/KaaDemo.cpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            if (leds[i].id == id) {
+                leds[i].status = status;
+                return;
+            }
+        }
+    }
+    
+    void printLedStatus(void)
+    {
+        printf("GPIO LED status: ");
```
Comment (RostakaGmfun) [inline gpiocontrol/source/cpp/src/KaaDemo.cpp]: Please prefer `std::cout`/`std::cin` over `printf`/`scanf`

File: gpiocontrol/source/cpp/src/KaaDemo.cpp [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    kaaClient->setEndpointAccessToken(DEMO_ACCESS_TOKEN);
+
+    /*
+     * Run the Kaa endpoint.
+     */
+    kaaClient->start();
+
+
+    while (true) {
+        std::this_thread::sleep_for(std::chrono::seconds(1));
```
Comment (RostakaGmfun) [inline gpiocontrol/source/cpp/src/KaaDemo.cpp:117]: You can omit sleep here. Infinite loop with empty body does the job well

File: gpiocontrol/source/cpp/src/KaaDemo.cpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    /*
+     * Stop the Kaa endpoint.
+     */
+    kaaClient->stop();
+
+    return 0;
```
Comment (RostakaGmfun) [inline gpiocontrol/source/cpp/src/KaaDemo.cpp]: Please use `EXIT_SUCCESS`

File: gpiocontrol/source/cpp/src/KaaDemo.cpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        for (std::size_t i = 0; i < leds.size(); ++i) {
+            if (leds[i].id == id) {
+                leds[i].status = status;
+                return;
+            }
+        }
+    }
+    
+    void printLedStatus(void)
```
Comment (RostakaGmfun) [inline gpiocontrol/source/cpp/src/KaaDemo.cpp]: `void` in place of argument list is not required in C++

**Comment by Slash32** [comment]
@kaa-leeroy add to whitelist

File: gpiocontrol/source/cpp/src/KaaDemo.cpp (hunk 4 above)
Comment (rasendubi) [reply gpiocontrol/source/cpp/src/KaaDemo.cpp:117]: Please, don't. Busy-looping wastes processor time, while this one does not.

File: gpiocontrol/source/cpp/src/KaaDemo.cpp (hunk 6 above)
Comment (vasner) [reply gpiocontrol/source/cpp/src/KaaDemo.cpp]: Fixed.

File: gpiocontrol/source/cpp/src/KaaDemo.cpp (hunk 5 above)
Comment (vasner) [reply gpiocontrol/source/cpp/src/KaaDemo.cpp]: Fixed.

File: gpiocontrol/source/cpp/src/KaaDemo.cpp (hunk 4 above)
Comment (vasner) [reply gpiocontrol/source/cpp/src/KaaDemo.cpp:117]: Ok!

File: gpiocontrol/source/cpp/src/KaaDemo.cpp (hunk 3 above)
Comment (vasner) [reply gpiocontrol/source/cpp/src/KaaDemo.cpp]: Fixed.

File: gpiocontrol/source/cpp/src/KaaDemo.cpp (hunk 1 above)
Comment (vasner) [reply gpiocontrol/source/cpp/src/KaaDemo.cpp:71]: Fixed.

File: gpiocontrol/source/cpp/src/KaaDemo.cpp (hunk 2 above)
Comment (vasner) [reply gpiocontrol/source/cpp/src/KaaDemo.cpp]: @RostakaGmfun, please, do not forget about, example :)

Comment (vasner) [reply gpiocontrol/source/cpp/src/KaaDemo.cpp]: Fixed.

File: gpiocontrol/resources/projects.xml [hunk 7]
```diff
@@ -159,6 +159,39 @@ After you successfully installed this app on your CC32XX device, go to "GPIO con
...
+        
+        <!-- Specifies the target runtime platforms according to the instructions from the project details section
+         (ex. LINUX_X86, ANDROID, IOS, ESP8266, etc., to get list of all possible options
+         see 'platform' type in demo_projects_schema.xsd schema) -->
+        
+        <platforms>LINUX_X86</platforms>
+        <features>EVENT USER_VERIFIER</features>
+        <complexity>REGULAR</complexity>
+        <bundleId>gpio_control</bundleId>
+        <sourceArchive>cpp-slave/gpiocontrol_demo.tar.gz/sourceArchive>
```
Comment (Slash32) [inline gpiocontrol/resources/projects.xml]: Please fix xml closing tag

**Review by Slash32 — CHANGES_REQUESTED** [review]

File: gpiocontrol/resources/projects.xml (hunk 7 above)
Comment (vasner) [reply gpiocontrol/resources/projects.xml]: Fixed.

File: gpiocontrol/assembly/src-cpp-slave.xml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    <id>gpiocontrol-src-cpp-slave</id>
+    <formats>
+        <format>tar.gz</format>
+    </formats>
+    <includeBaseDirectory>false</includeBaseDirectory>
+    <fileSets>
+        <fileSet>
+            <directory>${project.basedir}/source/cpp</directory>
+            <includes>
+                <include>build.sh</include>
```
Comment (Slash32) [inline gpiocontrol/assembly/src-cpp-slave.xml]: The build process is now moved to CMakeLists.txt use (See APP-505 and #564 ). The build.sh way is now obsolete and should not be used. Please update the build process accordngly.

File: gpiocontrol/resources/projects.xml [hunk 9]
```diff
@@ -159,6 +159,39 @@ After you successfully installed this app on your CC32XX device, go to "GPIO con
...
     </project>
+    <project id="gpiocontrol_demo_cpp_slave">
+        <name>GPIO control CPP</name>
+        <description>GPIO control CPP app that demonstrates capabilities of the Kaa Event subsystem</description>
+        <details>
+            <![CDATA[
+<h2>Installation</h2>
+Download the source code by clicking the "Source" button on the left. Unpack the downloaded archive and run the application using the following command in the console:
+<pre>
+$ ./build.sh
```
Comment (Slash32) [inline gpiocontrol/resources/projects.xml]: This should be also changed to CMakeList.txt

**Review by Slash32 — CHANGES_REQUESTED** [review]
The move to CMakeLists.txt is not fully completed. 

Also, the application configuration fails the Jenkins build.

Please update the PR accordingly.

File: gpiocontrol/assembly/src-cpp-slave.xml (hunk 8 above)
Comment (vasner) [reply gpiocontrol/assembly/src-cpp-slave.xml]: Fixed.

File: gpiocontrol/resources/projects.xml (hunk 9 above)
Comment (vasner) [reply gpiocontrol/resources/projects.xml]: Fixed.

**Comment by Slash32** [comment]
The PR has been re-based onto master as #575 
Also there was fixed gpiocontrol build

**Pull request closed without merging**